### PR TITLE
Add flag to forbid recovery in the parser

### PR DESCRIPTION
**Pull request by Noratrieb**
To start the effort of fixing #103534, this adds a new flag to the parser, which forbids the parser from doing recovery, which it shouldn't do in macros.

This doesn't add any new checks for recoveries yet and is just here to bikeshed the names for the functions here before doing more.

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
Let's do a perf run just to check that the parser size increase has no side-effects, otherwise we should probably just (mis/re)use the subparser name field as the "should recover" field.

@bors try @rust-timer queue

**Comment by Nilstrieb** [comment]
cc @estebank

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -491,6 +501,22 @@ impl<'a> Parser<'a> {
...
+        self.recovery = Recovery::Forbidden;
+        self
+    }
+
+    /// Whether the parser is allowed to recover from broken code.
+    ///
+    /// If this returns false, recovering broken code into valid code (especially if this recovery does lookahead)
+    /// is not allowed. All recovery done by the parser must be gated behind this check.
+    ///
+    /// Technically, this only needs to restruct eager recovery by doing lookahead at more tokens.
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/mod.rs:514]:
```suggestion
    /// Technically, this only needs to restrict eager recovery by doing lookahead at more tokens.
```

**Pull request merged**